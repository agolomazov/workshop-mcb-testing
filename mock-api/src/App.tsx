import { useQuery } from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import './App.css';

const axiosInstance = axios.create({
  baseURL: 'https://jsonplaceholder.typicode.com',
});

export interface User {
  id: number;
  name: string;
  username: string;
  email: string;
  address: Address;
  phone: string;
  website: string;
  company: Company;
}

interface Company {
  name: string;
  catchPhrase: string;
  bs: string;
}

interface Address {
  street: string;
  suite: string;
  city: string;
  zipcode: string;
  geo: Geo;
}

interface Geo {
  lat: string;
  lng: string;
}

function App() {
  const {
    data: users,
    error,
    isLoading,
  } = useQuery<User[], AxiosError>({
    queryKey: ['users'],
    queryFn: async () => {
      const response = axiosInstance.get('/users');
      return (await response).data;
    },
  });

  if (isLoading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>{error.message}</div>;
  }

  if (!users) {
    return <div>List users is empty...</div>;
  }

  return (
    <>
      <h1>Users List</h1>
      {users.map((user) => (
        <div key={user.id}>
          <p>
            Name: {user.name}
            <br />
            Email: <a href={`mailto:${user.email}`}>{user.email}</a>
            <br />
            Company: <b>{user.company.name}</b>
          </p>
        </div>
      ))}
    </>
  );
}

export default App;
