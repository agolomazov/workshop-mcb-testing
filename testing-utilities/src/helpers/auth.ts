import { isValidEmail, sendEmail } from './email';
import security from './security';

export async function signUp(email: string) {
  if (!isValidEmail(email)) return false;

  await sendEmail(email, 'Welcome aboard!');

  return true;
}

// Lesson: Spying on functions
export async function login(email: string) {
  const code = security.generateCode();

  await sendEmail(email, code.toString());
}
