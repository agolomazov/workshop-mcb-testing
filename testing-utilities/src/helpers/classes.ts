import {
  HUNDRED_PERCENT,
  LOW_PERCENT,
  MEDIUM_PERCENT,
} from '../constants/common';
import { ProgressColorClass } from '../types/common';

export function getProgressColorClass(percentage: number): ProgressColorClass {
  if (percentage < 0 || percentage > HUNDRED_PERCENT)
    throw new Error('invalid percentage value');
  if (percentage < LOW_PERCENT) return ProgressColorClass.RED;
  if (percentage < MEDIUM_PERCENT) return ProgressColorClass.YELLOW;
  if (percentage < HUNDRED_PERCENT) return ProgressColorClass.BLUE;

  return ProgressColorClass.GREEN;
}
