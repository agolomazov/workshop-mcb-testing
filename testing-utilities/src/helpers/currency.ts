export const getExchangeRate = (from: string, to: string) => {
  console.log(`Совершаем перевод средств из ${from} в ${to}`);
  return Math.random();
};
