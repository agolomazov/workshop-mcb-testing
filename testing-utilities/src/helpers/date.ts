import { MILLISECONDS_IN_SECOND } from '../constants/common';

export function formatSecondsWithSign(seconds: number): string {
  return `${seconds >= 0 ? '+' : '-'}${formatSeconds(seconds)}`;
}

export function formatSeconds(seconds: number): string {
  const date = new Date();

  date.setTime(Math.abs(seconds) * MILLISECONDS_IN_SECOND);

  const utc = date.toUTCString();

  return utc.substring(utc.indexOf(':') - 2, utc.indexOf(':') + 6);
}
