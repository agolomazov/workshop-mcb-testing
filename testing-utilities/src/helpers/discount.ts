export function getDiscount() {
  const today = new Date();
  const isChristmasDay = today.getMonth() === 11 && today.getDate() === 25;
  return isChristmasDay ? 0.2 : 0;
}
