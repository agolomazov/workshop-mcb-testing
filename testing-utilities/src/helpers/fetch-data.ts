export function fetchData() {
  return new Promise((resolve) => {
    setTimeout(() => {
      const data = [1, 2, 3];
      resolve(data);
    });
  });
}
