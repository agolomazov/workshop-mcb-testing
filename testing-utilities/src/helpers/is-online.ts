export function isOnline() {
  const availableHours = [8, 20];
  const [open, close] = availableHours;
  const currentHour = new Date().getHours();

  return currentHour >= open && currentHour < close;
}
