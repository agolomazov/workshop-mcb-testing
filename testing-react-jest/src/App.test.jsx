import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import App from './App';

import * as waitMock from './helpers/wait';

const waitSpy = jest.spyOn(waitMock, 'wait');
const promise = Promise.resolve();

describe('<App />', () => {
  const renderComponent = () => {
    const user = userEvent.setup();
    const { container } = render(<App />);

    const appEl = screen.getByTestId('app');
    const headingPage = screen.getByRole('heading');
    const formPage = screen.getByRole('form');
    const inputName = screen.getByRole('textbox', { name: /name/i });
    const inputPassword = screen.getByLabelText(/password/i);
    const submitBtn = screen.getByRole('button', { name: /create user/i });

    return {
      headingPage,
      formPage,
      user,
      inputName,
      inputPassword,
      container,
      submitBtn,
      appEl,
    };
  };

  it('Будет отображен заголовок и форма на странице', () => {
    const { headingPage, formPage, appEl } = renderComponent();

    expect(appEl).toBeInTheDocument();
    expect(headingPage).toBeInTheDocument();
    expect(headingPage.textContent).toMatch(/create user/i);
    expect(formPage).toBeInTheDocument();
  });

  it('Будет отображен errorMessage если мы нажимаем submit и вводим слабый пароль', async () => {
    const { user, submitBtn } = renderComponent();

    await user.click(submitBtn);

    const errorMessage = screen.queryByText(/8 characters long/i);

    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toHaveClass('error');
    expect(errorMessage.textContent).toMatchSnapshot();
  });

  it('Будет отображен successMessage если мы нажимаем submit и вводим валидные данные', async () => {
    waitSpy.mockImplementationOnce(() => promise);
    const { user, submitBtn, inputName, inputPassword } = renderComponent();

    await user.type(inputName, 'Anton');
    await user.type(inputPassword, '11aaAA##!23');
    await user.click(submitBtn);

    const successMessage = await screen.findByText(/created with password/i);

    expect(successMessage).toBeInTheDocument();
    expect(successMessage.textContent).toMatchSnapshot();
    expect(successMessage).toHaveClass('success');
  });
});
