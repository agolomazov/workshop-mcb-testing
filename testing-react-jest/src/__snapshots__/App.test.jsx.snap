// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Будет отображен errorMessage если мы нажимаем submit и вводим слабый пароль 1`] = `"Password must be at least 8 characters long"`;

exports[`<App /> Будет отображен successMessage если мы нажимаем submit и вводим валидные данные 1`] = `"User Anton created with password 11aaAA##!23"`;
